shadow-4.0.3/man/chsh.1
.\"$Id: chsh.1 6 2005-03-20 15:34:28Z bubulle $
.\" Copyright 1990, Julianne Frances Haugh
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Julianne F. Haugh nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH CHSH 1
.SH NAME
chsh \- change login shell
.SH SYNOPSIS
.TP 5
\fBchsh\fR
[\fB-s \fIlogin_shell\fR] [\fIuser\fR]
.SH DESCRIPTION
\fBchsh\fR changes the user login shell.
This determines the name of the user's initial login command.
A normal user may only change the login shell for her own account,
the super user may change the login shell for any account.
.PP
The only restrictions placed on the login shell is that the
command name must be listed in \fI/etc/shells\fR, unless the
invoker is the super-user, and then any value may be added.
An account with a restricted login shell may not change
her login shell.
For this reason, placing \fB/bin/rsh\fR in \fI/etc/shells\fR
is discouraged since accidentally changing to a restricted
shell would prevent the user from every changing her login
shell back to its original value.
.PP
If the \fB-s\fR option is not selected, \fBchsh\fR operates in an interactive
fashion, prompting the user with the current login shell.
Enter the new value to change the shell, or leave the line blank to use
the current one.
The current shell is displayed between a pair of \fB[ ]\fR marks.
.SH FILES
/etc/passwd \- user account information
.br
/etc/shells \- list of valid login shells
.SH SEE ALSO
.BR chfn (1),
.BR passwd (5)
.SH AUTHOR
Julianne Frances Haugh <jockgrrl@ix.netcom.com>


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



shadow-4.0.3/src/chsh.c
/*
 * Copyright 1989 - 1994, Julianne Frances Haugh
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <config.h>

#include "rcsid.h"
RCSID (PKG_VER "$Id: chsh.c 6 2005-03-20 15:34:28Z bubulle $")
#include <sys/types.h>
#include <stdio.h>
#include <fcntl.h>
#include <signal.h>
#include "prototypes.h"
#include "defines.h"
#include <pwd.h>
#include "pwio.h"
#include "getdef.h"
#include "pwauth.h"
#ifdef HAVE_SHADOW_H
#include <shadow.h>
#endif
#ifdef USE_PAM
#include "pam_defs.h"
#endif
#ifndef SHELLS_FILE
#define SHELLS_FILE "/etc/shells"
#endif
/* global variables */
static char *Prog;		/* Program name */
static int amroot;		/* Real UID is root */
static char loginsh[BUFSIZ];	/* Name of new login shell */

/* external identifiers */

#ifdef	NDBM
extern int pw_dbm_mode;
#endif

/* local function prototypes */
static void usage (void);
static void new_fields (void);
static int restricted_shell (const char *);

/*
 * usage - print command line syntax and exit
 */

static void usage (void)
{
	fprintf (stderr, _("Usage: %s [-s shell] [name]\n"), Prog);
	exit (1);
}

/*
 * new_fields - change the user's login shell information interactively
 *
 * prompt the user for the login shell and change it according to the
 * response, or leave it alone if nothing was entered.
 */

static void new_fields (void)
{
	printf (_
		("Enter the new value, or press return for the default\n"));
	change_field (loginsh, sizeof loginsh, _("Login Shell"));
}

/*
 * restricted_shell - return true if the named shell begins with 'r' or 'R'
 *
 * If the first letter of the filename is 'r' or 'R', the shell is
 * considered to be restricted.
 */

static int restricted_shell (const char *sh)
{
#if 0
	char *cp = Basename ((char *) sh);

	return *cp == 'r' || *cp == 'R';
#else
	/*
	 * Shells not listed in /etc/shells are considered to be restricted.
	 * Changed this to avoid confusion with "rc" (the plan9 shell - not
	 * restricted despite the name starting with 'r').  --marekm
	 */
	return !check_shell (sh);
#endif
}


/*
 * chsh - this command controls changes to the user's shell
 *
 *	The only supported option is -s which permits the the login shell to
 *	be set from the command line.
 */

int main (int argc, char **argv)
{
	char *user;		/* User name                         */
	int flag;		/* Current command line flag         */
	int sflg = 0;		/* -s - set shell from command line  */
	const struct passwd *pw;	/* Password entry from /etc/passwd   */
	struct passwd pwent;	/* New password entry                */

	sanitize_env ();

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	/*
	 * This command behaves different for root and non-root users.
	 */

	amroot = getuid () == 0;
#ifdef	NDBM
	pw_dbm_mode = O_RDWR;
#endif

	/*
	 * Get the program name. The program name is used as a prefix to
	 * most error messages.
	 */

	Prog = Basename (argv[0]);

	OPENLOG ("chsh");

	/*
	 * There is only one option, but use getopt() anyway to
	 * keep things consistent.
	 */

	while ((flag = getopt (argc, argv, "s:")) != EOF) {
		switch (flag) {
		case 's':
			sflg++;
			STRFCPY (loginsh, optarg);
			break;
		default:
			usage ();
		}
	}

	/*
	 * There should be only one remaining argument at most and it should
	 * be the user's name.
	 */

	if (argc > optind + 1)
		usage ();

	/*
	 * Get the name of the user to check. It is either the command line
	 * name, or the name getlogin() returns.
	 */

	if (optind < argc) {
		user = argv[optind];
		pw = getpwnam (user);
		if (!pw) {
			fprintf (stderr,
				 _("%s: Unknown user %s\n"), Prog, user);
			exit (1);
		}
	} else {
		pw = get_my_pwent ();
		if (!pw) {
			fprintf (stderr,
				 _
				 ("%s: Cannot determine your user name.\n"),
				 Prog);
			exit (1);
		}
		user = xstrdup (pw->pw_name);
	}

#ifdef	USE_NIS
	/*
	 * Now we make sure this is a LOCAL password entry for this user ...
	 */

	if (__ispwNIS ()) {
		char *nis_domain;
		char *nis_master;

		fprintf (stderr,
			 _("%s: cannot change user `%s' on NIS client.\n"),
			 Prog, user);

		if (!yp_get_default_domain (&nis_domain) &&
		    !yp_master (nis_domain, "passwd.byname",
				&nis_master)) {
			fprintf (stderr,
				 _
				 ("%s: `%s' is the NIS master for this client.\n"),
				 Prog, nis_master);
		}
		exit (1);
	}
#endif

	/*
	 * Non-privileged users are only allowed to change the shell if the
	 * UID of the user matches the current real UID.
	 */

	if (!amroot && pw->pw_uid != getuid ()) {
		SYSLOG ((LOG_WARN, "can't change shell for `%s'", user));
		closelog ();
		fprintf (stderr,
			 _("You may not change the shell for %s.\n"),
			 user);
		exit (1);
	}

	/*
	 * Non-privileged users are only allowed to change the shell if it
	 * is not a restricted one.
	 */

	if (!amroot && restricted_shell (pw->pw_shell)) {
		SYSLOG ((LOG_WARN, "can't change shell for `%s'", user));
		closelog ();
		fprintf (stderr,
			 _("You may not change the shell for %s.\n"),
			 user);
		exit (1);
	}

	/*
	   * Non-privileged users are optionally authenticated (must enter
	   * the password of the user whose information is being changed)
	   * before any changes can be made. Idea from util-linux
	   * chfn/chsh.  --marekm
	 */

	if (!amroot && getdef_bool ("CHFN_AUTH"))
		passwd_check (pw->pw_name, pw->pw_passwd, "chsh");

	/*
	 * Now get the login shell. Either get it from the password
	 * file, or use the value from the command line.
	 */

	if (!sflg)
		STRFCPY (loginsh, pw->pw_shell);

	/*
	 * If the login shell was not set on the command line, let the user
	 * interactively change it.
	 */

	if (!sflg) {
		printf (_("Changing the login shell for %s\n"), user);
		new_fields ();
	}

	/*
	 * Check all of the fields for valid information. The shell
	 * field may not contain any illegal characters. Non-privileged
	 * users are restricted to using the shells in /etc/shells.
	 * The shell must be executable by the user.
	 */

	if (valid_field (loginsh, ":,=")) {
		fprintf (stderr, _("%s: Invalid entry: %s\n"), Prog,
			 loginsh);
		closelog ();
		exit (1);
	}
	if (!amroot
	    && (!check_shell (loginsh) || access (loginsh, X_OK) != 0)) {
		fprintf (stderr, _("%s is an invalid shell.\n"), loginsh);
		closelog ();
		exit (1);
	}

	/*
	 * Before going any further, raise the ulimit to prevent
	 * colliding into a lowered ulimit, and set the real UID
	 * to root to protect against unexpected signals. Any
	 * keyboard signals are set to be ignored.
	 */

	if (setuid (0)) {
		SYSLOG ((LOG_ERR, "can't setuid(0)"));
		closelog ();
		fprintf (stderr, _("Cannot change ID to root.\n"));
		exit (1);
	}
	pwd_init ();

	/*
	 * The passwd entry is now ready to be committed back to
	 * the password file. Get a lock on the file and open it.
	 */

	if (!pw_lock ()) {
		SYSLOG ((LOG_WARN, "can't lock /etc/passwd"));
		closelog ();
		fprintf (stderr,
			 _
			 ("Cannot lock the password file; try again later.\n"));
		exit (1);
	}
	if (!pw_open (O_RDWR)) {
		SYSLOG ((LOG_ERR, "can't open /etc/passwd"));
		closelog ();
		fprintf (stderr, _("Cannot open the password file.\n"));
		pw_unlock ();
		exit (1);
	}

	/*
	 * Get the entry to update using pw_locate() - we want the real
	 * one from /etc/passwd, not the one from getpwnam() which could
	 * contain the shadow password if (despite the warnings) someone
	 * enables AUTOSHADOW (or SHADOW_COMPAT in libc).  --marekm
	 */
	pw = pw_locate (user);
	if (!pw) {
		pw_unlock ();
		fprintf (stderr,
			 _("%s: %s not found in /etc/passwd\n"), Prog,
			 user);
		exit (1);
	}

	/*
	 * Make a copy of the entry, then change the shell field. The other
	 * fields remain unchanged.
	 */
	pwent = *pw;
	pwent.pw_shell = loginsh;

	/*
	 * Update the passwd file entry. If there is a DBM file, update
	 * that entry as well.
	 */

	if (!pw_update (&pwent)) {
		SYSLOG ((LOG_ERR, "error updating passwd entry"));
		closelog ();
		fprintf (stderr,
			 _("Error updating the password entry.\n"));
		pw_unlock ();
		exit (1);
	}
#ifdef NDBM
	if (pw_dbm_present () && !pw_dbm_update (&pwent)) {
		SYSLOG ((LOG_ERR, "error updating DBM passwd entry"));
		closelog ();
		fprintf (stderr,
			 _("Error updating the DBM password entry.\n"));
		pw_unlock ();
		exit (1);
	}
	endpwent ();
#endif

	/*
	 * Changes have all been made, so commit them and unlock the file.
	 */

	if (!pw_close ()) {
		SYSLOG ((LOG_ERR, "can't rewrite /etc/passwd"));
		closelog ();
		fprintf (stderr,
			 _("Cannot commit password file changes.\n"));
		pw_unlock ();
		exit (1);
	}
	if (!pw_unlock ()) {
		SYSLOG ((LOG_ERR, "can't unlock /etc/passwd"));
		closelog ();
		fprintf (stderr, _("Cannot unlock the password file.\n"));
		exit (1);
	}
	SYSLOG ((LOG_INFO, "changed user `%s' shell to `%s'", user,
		 loginsh));
	closelog ();
	exit (0);
}
